core/blockchain_mmr.go

```diff
@@ -0,0 +1,242 @@
+package core
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/harmony-one/harmony/block"
+	"github.com/harmony-one/harmony/internal/mmr"
+	"github.com/harmony-one/harmony/internal/mmr/db"
+	"github.com/harmony-one/harmony/shard"
+
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+)
+
+type OpenMmrDB func(epoch *big.Int, create bool) *mmr.Mmr
+
+type BlockchainMMR struct {
+	bc            *BlockChain
+	cache         map[uint64]*mmr.Mmr
+	cacheLock     sync.RWMutex
+	MemMmrDB      *mmr.Mmr
+	MmrDBLock     sync.RWMutex
+	MmrCommitChan chan *mmr.Mmr
+	quit          chan struct{}
+	wait          sync.WaitGroup
+}
+
+func newMemMmrDB(epoch *big.Int) *mmr.Mmr {
+	leafLength := shard.Schedule.InstanceForEpoch(epoch).BlocksPerEpoch()
```

Comment:
does our MMR logic support the change of leafLength? what happens if the blocksPerEpoch change in the future?

Reply:
yeah, MMR supports arbitrary leaf length. here `leafLength` is just used to optimize map memory allocation.

Reply:
Will the (disk) database need to be re-created in the event the `leafLength` changes?

Reply:
> Will the (disk) database need to be re-created in the event the `leafLength` changes?

don't need.